cpp/include/ucxx/delayed_submission.h

```diff
@@ -122,20 +136,47 @@ class BaseDelayedSubmissionCollection {
    */
   void process()
   {
-    decltype(_collection) itemsToProcess;
+    // Process only those that were already inserted to prevent from never
+    // returning if `_collection` grows indefinitely.
+    size_t toProcess = 0;
     {
       std::lock_guard<std::mutex> lock(_mutex);
-      // Move _collection to a local copy in order to to hold the lock for as
-      // short as possible
-      itemsToProcess = std::move(_collection);
+      toProcess = _collection.size();
     }
 
-    if (itemsToProcess.size() > 0) {
-      ucxx_trace_req("Submitting %lu %s callbacks", itemsToProcess.size(), _name.c_str());
-      for (auto& item : itemsToProcess)
-        processItem(item);
+    for (auto i = 0; i < toProcess; ++i) {
+      std::pair<ItemIdType, T> item;
+      {
+        std::lock_guard<std::mutex> lock(_mutex);
```

Comment:
The worry is it may take too long, particularly critical if something takes too long on the UCX end to complete. Most of the tasks we use are supposed to be non-blocking, but I've encountered weird issues in very specific scenarios that may occur with one transport but not another. The problem if this occurs is we block the main thread if it's trying to schedule/cancel another request. It may be a bit overreaching and perhaps we could be able to indeed do as you say, but I'd prefer to do it carefully in the future, I'd hate to fix the segfault just to introduce a new potential deadlock here.